### PR TITLE
[MXNET-1224]: improve scala maven jni build.

### PR DESCRIPTION
**Pull request by frankfliu**
## Description ##
Scala jni shared library requires compile flags inherited from make. But there is no way to enforce those flags to be the same as libmxnet.so was built. And current Scala package is built is manually. This is error prone can has caused issues in past. 

This fix is trying make scala jni library not depends on any mxnet build flags, and dynamically link libmxnet.so directly. libmxnet-scala.so file will be compiled dynamically link to libmxnet.so, and allows load libmxnet.so from the same directory. Scala code is also updated to extract libmxnet.so file into temp folder and load it temp folder.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by nswamy** [comment]
@frankfliu could you please explain what you intend to do.

**Comment by vandanavk** [comment]
@mxnet-label-bot add [Scala, pr-work-in-progress]

File: scala-package/init-native/linux-x86_64/pom.xml [hunk 1]
```diff
@@ -62,22 +66,24 @@
             <compilerStartOption>-std=c++0x</compilerStartOption>
           </compilerStartOptions>
           <compilerEndOptions>
-            <compilerEndOption>-I${project.basedir}/../../../include</compilerEndOption>
-            <compilerEndOption>${all_includes}</compilerEndOption>
-            <compilerEndOption>${cflags}</compilerEndOption>
+            <compilerEndOption>-I${MXNET_DIR}/include</compilerEndOption>
```
Comment (nswamy) [inline scala-package/init-native/linux-x86_64/pom.xml]: Can we use the C_FLAGS and LD_FLAGS defined and passed from the Makefile

File: scala-package/pom.xml [hunk 2]
```diff
@@ -39,6 +39,8 @@
     <scala.version>2.11.8</scala.version>
     <scala.binary.version>2.11</scala.binary.version>
     <build.platform />
+    <cxx>g++</cxx>
+    <dollor>$</dollor>
```
Comment (zachgk) [inline scala-package/pom.xml]: Should this be dollar?

Comment (frankfliu) [reply scala-package/pom.xml]: True, will fix this in next PR with maven build improvement.

File: scala-package/init-native/osx-x86_64/pom.xml [hunk 3]
```diff
@@ -62,8 +66,14 @@
...
           <compilerEndOptions>
-            <compilerEndOption>-I${project.basedir}/../../../include</compilerEndOption>
-            <compilerEndOption>${cflags}</compilerEndOption>
+            <compilerEndOption>-I${MXNET_DIR}/include</compilerEndOption>
+            <compilerEndOption>-I${MXNET_DIR}/3rdparty/dmlc-core/include</compilerEndOption>
+            <compilerEndOption>-I${MXNET_DIR}/3rdparty/mshadow</compilerEndOption>
+            <compilerEndOption>-I${MXNET_DIR}/3rdparty/dlpack/include</compilerEndOption>
+            <compilerEndOption>-I${MXNET_DIR}/3rdparty/tvm/nnvm/include</compilerEndOption>
+            <compilerEndOption>-DMSHADOW_USE_MKL=0 -DMSHADOW_USE_CUDA=0</compilerEndOption>
+            <compilerEndOption>-g -O0 -fPIC -msse3 -mf16c</compilerEndOption>
```
Comment (lanking520) [inline scala-package/init-native/osx-x86_64/pom.xml]: Why we are using these build options? What are these for?

File: scala-package/init-native/linux-x86_64/pom.xml [hunk 4]
```diff
@@ -62,22 +66,24 @@
...
+            <compilerEndOption>-I${MXNET_DIR}/3rdparty/dmlc-core/include</compilerEndOption>
+            <compilerEndOption>-I${MXNET_DIR}/3rdparty/mshadow</compilerEndOption>
+            <compilerEndOption>-I${MXNET_DIR}/3rdparty/dlpack/include</compilerEndOption>
+            <compilerEndOption>-I${MXNET_DIR}/3rdparty/tvm/nnvm/include</compilerEndOption>
+            <compilerEndOption>-DMSHADOW_USE_MKL=0 -DMSHADOW_USE_CUDA=0</compilerEndOption>
```
Comment (lanking520) [inline scala-package/init-native/linux-x86_64/pom.xml]: Why we make them =0, if we build with GPU what will happened?

**Comment by lanking520** [comment]
@marcoabreu looks like the CI passed however not reflect on this PR?

File: scala-package/init-native/linux-x86_64/pom.xml (hunk 4 above)
Comment (frankfliu) [reply scala-package/init-native/linux-x86_64/pom.xml]: Added a README.md file to describe compiler and linker flags.

**Pull request merged**